### PR TITLE
Add LiteX DNA driver

### PR DESCRIPTION
**Pull request by mateusz-holenko**
This adds LiteX DNA ID driver and enables it in litex_vexriscv board.

### CODE REVIEW

File: drivers/hwinfo/hwinfo_litex.c
Comment (alexanderwachter) [inline drivers/hwinfo/hwinfo_litex.c]: You should return end here.

File: drivers/hwinfo/hwinfo_litex.c
Comment (alexanderwachter) [inline drivers/hwinfo/hwinfo_litex.c]: Can you point me to the documentation where I can find some information about this?

**Review by alexanderwachter — CHANGES_REQUESTED** [review]

File: drivers/hwinfo/hwinfo_litex.c
Comment (mateusz-holenko) [reply drivers/hwinfo/hwinfo_litex.c]: Of course, my bad.

File: drivers/hwinfo/hwinfo_litex.c
Comment (mateusz-holenko) [reply drivers/hwinfo/hwinfo_litex.c]: I believe that in case of LiteX, the source code together with internal comments *is* the documentation. 

The DNA's code is here: https://github.com/enjoy-digital/litex/blob/master/litex/soc/cores/dna.py, but you are most probably interested in the CSRStatus: https://github.com/enjoy-digital/litex/blob/master/litex/soc/interconnect/csr.py#L125.

Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_litex.c]: I was asking for documentation about this DNA module — something about the content of this constants.
The comment from the DT defines this register as 
"DNA value is 47-bits long, register width is aligned to 64",
But here you have ` *(ptr + (i * 0x04));` where `i` is maximum 8 (DT_INST_0_LITEX_DNA0_SIZE / 8).
`8 * 0x04` evaluates to 32. That seems like an overrun, right?
The code is the same as:
```C
u32_t volatile *ptr = (u32_t volatile *)(DT_INST_0_LITEX_DNA0_BASE_ADDRESS);
ssize_t end = MIN(length, (DT_INST_0_LITEX_DNA0_SIZE / 8));

for (int i = 0; i < end; i++) {
	buffer[i] = ptr[i] & 0x08; /* In each CSR only lower 8 bits are used */
}
```

Comment (mateusz-holenko) [reply drivers/hwinfo/hwinfo_litex.c]: I found another source that might shed some light on how CSRs are defined in LiteX: https://github.com/timvideos/litex-buildenv/wiki/LiteX-for-Hardware-Engineers#csrs-config-and-status-registers. The most important part is:
```
CSRs are a bit odd, by default they are byte-wide registers that are
on 32-bit word boundaries.
So a "32-bit" CSR is actually broken into four bytes spanning a total
address space of 16 bytes.
```
I've just noticed there is a typo in the comment that is quite importaint in the discussion -- DNA ID has **57** bits instead of 47. It means that 8 bytes are required to represent it (as 57 = 7*8 + 1).

As it does not fit in a single 32-bit register, it is exposed as a 64-bit CSR (which can be treated as two 32-bit CSRs). Since LiteX generates a single 32-bit CSR as 4 consecutive 32-bit registers, there are 8 registers in total (each one containing meaningful data stored in the least-significant byte).

I'll fix the typo and rephrase the comment in the DTS to make things clearer.

Comment (mateusz-holenko) [reply drivers/hwinfo/hwinfo_litex.c]: 57 comes from: https://github.com/enjoy-digital/litex/blob/master/litex/soc/cores/dna.py#L11  which corresponds to the info on `DNA_PORT` on page 111 in this [document](https://www.xilinx.com/support/documentation/sw_manuals/xilinx14_7/7series_hdl.pdf).

Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_litex.c]: Can we use the code from my snipped above? I think that notation is somehow more clear.

File: dts/riscv/riscv32-litex-vexriscv.dtsi
Comment (alexanderwachter) [inline dts/riscv/riscv32-litex-vexriscv.dtsi]: I think we usually use the numbers of registers here. Not the numbers of bits. But maybe I'm wrong.

File: drivers/hwinfo/hwinfo_litex.c
Comment (mateusz-holenko) [reply drivers/hwinfo/hwinfo_litex.c]: Yes. I am reprhasing the comment in the code as well and I think it will fit better in the context of your notation.
I think thougt, that it should be: `buffer[i] = ptr[i] & 0xFF`. Plus I would add casting.

File: dts/riscv/riscv32-litex-vexriscv.dtsi
Comment (mateusz-holenko) [reply dts/riscv/riscv32-litex-vexriscv.dtsi]: I don't know if there is a strict definition of what the size means in Zephyr.
From what I see in other sections of this and other dtses it looks that the value should be the sum of widths of all registers expressed in *bytes*, so in this context: 0x20 (8 registers, each 4-byte).

I'll change it to be consistent and avoid further confusion.

File: drivers/hwinfo/hwinfo_litex.c
Comment (alexanderwachter) [inline drivers/hwinfo/hwinfo_litex.c]: Just a little nick-pit:
use `sizeof(u32_t)` instead of the constant 4.

Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_litex.c]: Otherwise, it looks good for me. Unfortunately, I don't have any FPGA boards to test.
Did you run the test-cases? There is a shell module to verify the output. Please do so if you haven't already.

File: drivers/hwinfo/hwinfo_litex.c
Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_litex.c]: You are right. Don't know why I wrote 0x08. An explicit cast is always a good idea.

File: drivers/hwinfo/hwinfo_litex.c
Comment (mateusz-holenko) [reply drivers/hwinfo/hwinfo_litex.c]: > Just a little nick-pit:
> use `sizeof(u32_t)` instead of the constant 4.

Changed.

> Otherwise, it looks good for me. Unfortunately, I don't have any FPGA boards to test.
> Did you run the test-cases? There is a shell module to verify the output. Please do so if you haven't already.

I've tested it with `shell_module` demo with `hwinfo_shell` on Arty board on two pieces of HW:
```
uart:~$ hwinfo devid
Length: 8
ID: 0x005c1d2e50210854
```

```
uart:~$ hwinfo devid
Length: 8
ID: 0x005cc54e233ef054
```

**Review by alexanderwachter — APPROVED** [review]

**Comment by mateusz-holenko** [comment]
@alexanderwachter The code is rebased on top of the 2.0.0 release. Is there anything else to do here that blocks merging the PR?

**Comment by alexanderwachter** [comment]
@mateusz-holenko no. Just wait until someone with merge rights merges it. Ping @galak

**Review by carlescufi — APPROVED** [review]

**Comment by mateusz-holenko** [comment]
Thanks!

**Pull request merged**
